File: java/Figuras/src/main/java/org/example/FormateadorSalida.java
package org.example;

import java.util.List;
import java.util.Map;

/**
 * Clase para formatear la salida hacia el usuario
 */
public class FormateadorSalida {
    
    /**
     * Formatea un menú para mostrar al usuario
     * @param titulo Título del menú
     * @param opciones Lista de opciones del menú
     * @param info Información adicional (opcional)
     * @return Menú formateado
     */
    public static String formatearMenu(String titulo, List<String> opciones, String info) {
        StringBuilder resultado = new StringBuilder();
        String lineaSeparadora = "=".repeat(50);
        
        resultado.append(lineaSeparadora).append("\n");
        resultado.append(String.format(" %s \n", centrarTexto(titulo, 48)));
        resultado.append(lineaSeparadora).append("\n");
        
        if (info != null && !info.isEmpty()) {
            resultado.append(info).append("\n");
            resultado.append("\n");
        }
        
        for (int i = 0; i < opciones.size(); i++) {
            resultado.append(String.format("%2d. %s\n", i + 1, opciones.get(i)));
        }
        
        resultado.append(lineaSeparadora);
        return resultado.toString();
    }
    
    /**
     * Formatea la información de una figura
     * @param figura Figura a formatear
     * @param unidad Unidad de medida a usar
     * @return Información formateada de la figura
     */
    public static String formatearFigura(Figura figura, UnidadMedida unidad) {
        StringBuilder resultado = new StringBuilder();
        
        resultado.append("📐 Figura ID: ").append(figura.getId()).append("\n");
        resultado.append("   Tipo: ").append(figura.getNombre()).append("\n");
        resultado.append("   Categoría: ").append(figura.getTipo()).append("\n");
        
        // Mostrar dimensiones específicas
        if (figura instanceof Circulo) {
            Circulo circulo = (Circulo) figura;
            double radioConvertido = UnidadAdapter.convertir(circulo.getRadio(), UnidadMedida.METROS, unidad);
            resultado.append("   Radio: ").append(UnidadAdapter.formatearConUnidad(radioConvertido, unidad)).append("\n");
        } else if (figura instanceof Cuadrado) {
            Cuadrado cuadrado = (Cuadrado) figura;
            double ladoConvertido = UnidadAdapter.convertir(cuadrado.getLado(), UnidadMedida.METROS, unidad);
            resultado.append("   Lado: ").append(UnidadAdapter.formatearConUnidad(ladoConvertido, unidad)).append("\n");
        } else if (figura instanceof Cubo) {
            Cubo cubo = (Cubo) figura;
            double ladoConvertido = UnidadAdapter.convertir(cubo.getLado(), UnidadMedida.METROS, unidad);
            resultado.append("   Lado: ").append(UnidadAdapter.formatearConUnidad(ladoConvertido, unidad)).append("\n");
        } else if (figura instanceof Esfera) {
            Esfera esfera = (Esfera) figura;
            double radioConvertido = UnidadAdapter.convertir(esfera.getRadio(), UnidadMedida.METROS, unidad);
            resultado.append("   Radio: ").append(UnidadAdapter.formatearConUnidad(radioConvertido, unidad)).append("\n");
        }
        
        // Mostrar cálculos
        if (figura instanceof Figura2d) {
            Figura2d figura2d = (Figura2d) figura;
            double area = figura2d.calcularArea();
            double perimetro = figura2d.calcularPerimetro();
            
            // Convertir área (unidad²) y perímetro
            double factor = unidad.getFactorConversion();
            double areaConvertida = area / (factor * factor);
            double perimetroConvertido = UnidadAdapter.convertir(perimetro, UnidadMedida.METROS, unidad);
            
            resultado.append(String.format("   Área: %.2f %s²\n", areaConvertida, unidad.getSimbolo()));
            resultado.append("   Perímetro: ").append(UnidadAdapter.formatearConUnidad(perimetroConvertido, unidad)).append("\n");
            
        } else if (figura instanceof Figura3d) {
            Figura3d figura3d = (Figura3d) figura;
            double volumen = figura3d.calcularVolumen();
            
            // Convertir volumen (unidad³)
            double factor = unidad.getFactorConversion();
            double volumenConvertido = volumen / (factor * factor * factor);
            
            resultado.append(String.format("   Volumen: %.2f %s³\n", volumenConvertido, unidad.getSimbolo()));
            
            if (figura instanceof Cubo) {
                Cubo cubo = (Cubo) figura;
                double areaSuperficie = cubo.calcularAreaSuperficie();
                double areaConvertida = areaSuperficie / (factor * factor);
                resultado.append(String.format("   Área superficie: %.2f %s²\n", areaConvertida, unidad.getSimbolo()));
            } else if (figura instanceof Esfera) {
                Esfera esfera = (Esfera) figura;
                double areaSuperficie = esfera.calcularAreaSuperficie();
                double areaConvertida = areaSuperficie / (factor * factor);
                resultado.append(String.format("   Área superficie: %.2f %s²\n", areaConvertida, unidad.getSimbolo()));
            }
        }
        
        return resultado.toString();
    }
    
    /**
     * Formatea una lista de figuras
     * @param figuras Lista de figuras
     * @param unidad Unidad de medida a usar
     * @return Lista formateada de figuras
     */
    public static String formatearListaFiguras(List<Figura> figuras, UnidadMedida unidad) {
        if (figuras.isEmpty()) {
            return "No hay figuras para mostrar.";
        }
        
        StringBuilder resultado = new StringBuilder();
        resultado.append("Total de figuras: ").append(figuras.size()).append("\n");
        resultado.append("=".repeat(50)).append("\n");
        
        for (int i = 0; i < figuras.size(); i++) {
            resultado.append("\n").append(i + 1).append(". ").append(formatearFigura(figuras.get(i), unidad));
        }
        
        return resultado.toString();
    }
    
    /**
     * Formatea las estadísticas del repositorio
     * @param estadisticas Mapa con estadísticas
     * @return Estadísticas formateadas
     */
    public static String formatearEstadisticas(Map<String, Integer> estadisticas) {
        StringBuilder resultado = new StringBuilder();
        resultado.append("ESTADÍSTICAS DEL REPOSITORIO\n");
        resultado.append("=".repeat(40)).append("\n");
        resultado.append("Total de figuras: ").append(estadisticas.getOrDefault("total", 0)).append("\n");
        resultado.append("\n");
        resultado.append("Distribución por tipo:\n");
        
        String[] tipos = {"círculo", "cuadrado", "cubo", "esfera"};
        int total = estadisticas.getOrDefault("total", 0);
        
        for (String tipo : tipos) {
            int cantidad = estadisticas.getOrDefault(tipo, 0);
            if (cantidad > 0) {
                double porcentaje = total > 0 ? (cantidad * 100.0 / total) : 0.0;
                resultado.append(String.format(" • %s: %d (%.1f%%)\n", 
                    tipo.substring(0, 1).toUpperCase() + tipo.substring(1), cantidad, porcentaje));
            }
        }
        
        return resultado.toString();
    }
    
    /**
     * Formatea un mensaje de error
     * @param mensaje Mensaje de error
     * @return Mensaje de error formateado
     */
    public static String mostrarError(String mensaje) {
        return "Error: " + mensaje;
    }
    
    /**
     * Formatea un mensaje de éxito
     * @param mensaje Mensaje de éxito
     * @return Mensaje de éxito formateado
     */
    public static String mostrarExito(String mensaje) {
        return mensaje;
    }
    
    /**
     * Formatea un mensaje informativo
     * @param mensaje Mensaje informativo
     * @return Mensaje informativo formateado
     */
    public static String mostrarInfo(String mensaje) {
        return mensaje;
    }
    
    /**
     * Centra un texto en un ancho específico
     * @param texto Texto a centrar
     * @param ancho Ancho total
     * @return Texto centrado
     */
    private static String centrarTexto(String texto, int ancho) {
        if (texto.length() >= ancho) {
            return texto;
        }
        
        int espacios = (ancho - texto.length()) / 2;
        return " ".repeat(espacios) + texto + " ".repeat(ancho - texto.length() - espacios);
    }
}

File: python/Figuras/FormateadorSalida.py
"""
Clase para formatear la salida hacia el usuario
"""
from typing import List, Dict, Any
from .Figura import Figura
from .Figura2d import Figura2d
from .Figura3d import Figura3d
from .UnidadMedida import UnidadMedida
from .UnidadAdapter import UnidadAdapter


class FormateadorSalida:
    """Clase para formatear y presentar información al usuario"""
    
    @staticmethod
    def formatear_menu(titulo: str, opciones: List[str], info: str = None) -> str:
        """
        Formatea un menú para mostrar al usuario
        
        Args:
            titulo (str): Título del menú
            opciones (List[str]): Lista de opciones del menú
            info (str): Información adicional (opcional)
            
        Returns:
            str: Menú formateado
        """
        linea_separadora = "=" * 50
        resultado = [linea_separadora]
        resultado.append(f" {titulo.center(48)} ")
        resultado.append(linea_separadora)
        
        if info:
            resultado.append(f"ℹ️  {info}")
            resultado.append("")
        
        for i, opcion in enumerate(opciones, 1):
            resultado.append(f"{i:2d}. {opcion}")
        
        resultado.append(linea_separadora)
        
        return "\n".join(resultado)
    
    @staticmethod
    def formatear_figura(figura: Figura, unidad: UnidadMedida) -> str:
        """
        Formatea la información de una figura
        
        Args:
            figura (Figura): Figura a formatear
            unidad (UnidadMedida): Unidad de medida a usar
            
        Returns:
            str: Información formateada de la figura
        """
        resultado = []
        resultado.append(f"📐 Figura ID: {figura.get_id()}")
        resultado.append(f"   Tipo: {figura.get_nombre()}")
        resultado.append(f"   Categoría: {figura.get_tipo()}")
        
        # Mostrar dimensiones específicas
        if hasattr(figura, 'get_radio'):
            radio_convertido = UnidadAdapter.convertir(figura.get_radio(), UnidadMedida.METROS, unidad)
            resultado.append(f"   Radio: {UnidadAdapter.formatear_con_unidad(radio_convertido, unidad)}")
        elif hasattr(figura, 'get_lado'):
            lado_convertido = UnidadAdapter.convertir(figura.get_lado(), UnidadMedida.METROS, unidad)
            resultado.append(f"   Lado: {UnidadAdapter.formatear_con_unidad(lado_convertido, unidad)}")
        
        # Mostrar cálculos
        if isinstance(figura, Figura2d):
            area = figura.calcular_area()
            perimetro = figura.calcular_perimetro()
            
            # Convertir área (unidad²) y perímetro
            factor = unidad.get_factor_conversion()
            area_convertida = area / (factor ** 2)
            perimetro_convertido = UnidadAdapter.convertir(perimetro, UnidadMedida.METROS, unidad)
            
            resultado.append(f"   Área: {area_convertida:.2f} {unidad.get_simbolo()}²")
            resultado.append(f"   Perímetro: {UnidadAdapter.formatear_con_unidad(perimetro_convertido, unidad)}")
        
        elif isinstance(figura, Figura3d):
            volumen = figura.calcular_volumen()
            
            # Convertir volumen (unidad³)
            factor = unidad.get_factor_conversion()
            volumen_convertido = volumen / (factor ** 3)
            
            resultado.append(f"   Volumen: {volumen_convertido:.2f} {unidad.get_simbolo()}³")
            
            if hasattr(figura, 'calcular_area_superficie'):
                area_superficie = figura.calcular_area_superficie()
                area_convertida = area_superficie / (factor ** 2)
                resultado.append(f"   Área superficie: {area_convertida:.2f} {unidad.get_simbolo()}²")
        
        return "\n".join(resultado)
    
    @staticmethod
    def formatear_lista_figuras(figuras: List[Figura], unidad: UnidadMedida) -> str:
        """
        Formatea una lista de figuras
        
        Args:
            figuras (List[Figura]): Lista de figuras
            unidad (UnidadMedida): Unidad de medida a usar
            
        Returns:
            str: Lista formateada de figuras
        """
        if not figuras:
            return "📭 No hay figuras para mostrar."
        
        resultado = []
        resultado.append(f"📊 Total de figuras: {len(figuras)}")
        resultado.append("=" * 50)
        
        for i, figura in enumerate(figuras, 1):
            resultado.append(f"\n{i}. {FormateadorSalida.formatear_figura(figura, unidad)}")
        
        return "\n".join(resultado)
    
    @staticmethod
    def formatear_estadisticas(estadisticas: Dict[str, Any]) -> str:
        """
        Formatea las estadísticas del repositorio
        
        Args:
            estadisticas (Dict[str, Any]): Diccionario con estadísticas
            
        Returns:
            str: Estadísticas formateadas
        """
        resultado = []
        resultado.append("📈 ESTADÍSTICAS DEL REPOSITORIO")
        resultado.append("=" * 40)
        resultado.append(f"Total de figuras: {estadisticas.get('total', 0)}")
        resultado.append("")
        resultado.append("Distribución por tipo:")
        
        tipos = ['círculo', 'cuadrado', 'cubo', 'esfera']
        for tipo in tipos:
            cantidad = estadisticas.get(tipo, 0)
            if cantidad > 0:
                porcentaje = (cantidad / estadisticas['total']) * 100 if estadisticas['total'] > 0 else 0
                resultado.append(f"  • {tipo.capitalize()}: {cantidad} ({porcentaje:.1f}%)")
        
        return "\n".join(resultado)
    
    @staticmethod
    def mostrar_error(mensaje: str) -> str:
        """
        Formatea un mensaje de error
        
        Args:
            mensaje (str): Mensaje de error
            
        Returns:
            str: Mensaje de error formateado
        """
        return f"❌ Error: {mensaje}"
    
    @staticmethod
    def mostrar_exito(mensaje: str) -> str:
        """
        Formatea un mensaje de éxito
        
        Args:
            mensaje (str): Mensaje de éxito
            
        Returns:
            str: Mensaje de éxito formateado
        """
        return f"✅ {mensaje}"
    
    @staticmethod
    def mostrar_info(mensaje: str) -> str:
        """
        Formatea un mensaje informativo
        
        Args:
            mensaje (str): Mensaje informativo
            
        Returns:
            str: Mensaje informativo formateado
        """
        return f"ℹ️  {mensaje}"